app/lua/lauxlib.c

```diff
@@ -791,6 +792,11 @@ static void *l_alloc (void *ud, void *ptr, size_t osize, size_t nsize) {
   }
   if (L != NULL && (mode & EGC_ALWAYS)) /* always collect memory if requested */
     luaC_fullgc(L);
+#ifndef LUA_CROSS_COMPILER
+  if (L != NULL && (mode & EGC_ON_MEM_LIMIT) && G(L)->memlimit < 0 &&
+      (system_get_free_heap_size() < (-G(L)->memlimit)))
```

Comment:
I was going to ask expensive in terms of clock cycles is the `system_get_free_heap_size()` until I realised that I should answer this myself since I've got a remote debug session onto a running firmware.  The call is a wrapper around a getter in `libmain.a(mem_manager.o)` and the value itself is maintained by the memory manager in the 0th word of its `.bss`, so th tl;dr is that the call it dirt cheap.